# Практическая работа №0: Исследование алгоритмов формирования аддитивных цепочек

Выполнила студентка гр. 1304 Ярусова Татьяна. Вариант №53.

## Цель работы

Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных аддитивных цепочек для заданного числа, привить навык использования систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения

#### Аддитивные цепочки.<br>
**Определение.**<br>
Аддитивной цепочкой для натурального $n$ называется последовательность натуральных чисел<br>
$$ 1 = a_0, a_1, a_2,\dots,a_r = n, $$
где каждый элемент последовательности равен сумме каких-либо двух предыдущих элементов:
$$ a_i = a_j + a_k,\quad k \leq j < i,\quad i = 1, 2, \dots, r. $$

**Связь аддитивных цепочек с методами эффективного возведения мономов в степень.**<br>
<ul>
    <li>$l(n) = r$ - наименьшая длина $r$, для которой существует аддитивная цепочка для значения $n$ (функция от $n$).</li>
    <li>Неравенство для метода множителей:</li>
        $$ l(mn) \leq l(m) + l(n)$$
    <li>Неравенство $m$-арного метода ($m = 2^k, n = \sum\limits_{j=0}^td_jm^{t-j}$):</li>
        $$ l(n) \leq m - 2 + (k + 1)^t.$$
    <li>Неравенство для бинарного метода "SX":</li>
        $$ l(n) \leq \lambda(n) + \nu(n) - 1. $$
</ul>

**Свойства аддитивных цепочек.**<br>
**Шаг аддитивной цепочки.**<br>
<ul>
    <li>Полагаем, что все аддитивные цепочки возрастающие:</li>
        $$ 1 = a_0 < a_1 < a_2 < \dots < a_r = n. $$
    <li>Если два числа из $\{a_i\}$ одинаковые, то одно из них может быть опущено.</li>
    <li>Пара $(j, k), \; 0 \leq k \leq j < i$ - называется шагом.</li>
    <li>Если существует более чем одна пара $(j, k)$, полагаем, что $j$ - наибольшее из возможных.</li>
    </ul>
    
**Виды шагов.**
<ul>
    <li>Удвоение:</li>
        $ j = k = i - 1 $
    <li>Звёздный:</li>
        $ j = i - 1 $
    <li>Малый:</li>
        $ \lambda(a_i) = \lambda(a_{i-1}) $
    </ul>
    
**Свойства шагов.**
<ul>
    <li>Первый шаг всегда удвоение.</li>
    <li>Удвоение - звёздный шаг, но никогда не малый шаг.</li>
    <li>За удвоением всегда следует звёздный шаг.</li>
    <li>Если $i$-ый шаг не малый, то $(i+1)$ шаг либо малый, либо звёздный, либо и тот, и другой.</li>
    <li>Если $(i+1)$ шаг ни звездный, ни малый, то $i$-ый шаг должен быть малым.</li>
    </ul>
 

#### Приближённые алгоритмы нахождения аддитивных цепочек.<br>
**Алгоритм Брауэра.**<br>
<ul>
    <li>Алгоритм Брауэра позволяет вычислить $n$ степень за количество умножений равное</li>
    $$ l_b(n) = \lambda(n) + \frac{\lambda(n)}{\lambda(\lambda(n))} + O\left(\frac{\lambda(n)\cdot\lambda(\lambda(\lambda(n)))}{(\lambda(\lambda(n)))^2}\right) $$
    <li>Самая короткая аддитивная цепчока для числа $n$ имеет длину не более $\lambda(n)$.</li>
    <li>Доказано, что почти для всех $n$ минимальная аддитивная цепчока имеет длину $l_b(n)$.</li>
    </ul>
    
**Реккурентное определение.**<br>
Для наутральных $n$ при заданном натуральном числе $k$ можно построить цепочку Брауэра с помощью реккурентной формулы:
$$  B_k(n) =   \begin{cases}
1, 2, 3,\dots, 2^k-1       & \quad \text{если } n < 2^k\\
B_k(q), 2q, 4q, 8q,\dots, 2^kq, n  & \quad \text{если } n \geqslant 2^k \text{ и } q = \lfloor\frac{n}{2^k}\rfloor\end{cases} $$<br>
Данная цепочка имеет длину:<br>
$$ l_b(n) = j(k+1) + 2^k - 2 $$<br>
при условии, что $jk \leq \lambda(n) < (j+1)k.$<br>
Длина цепочки будет минимизирована для больших $n$, если<br>
$$ k = \lambda(\lambda(n)) - 2\lambda(\lambda(\lambda(n))) $$

**Уточнение алгоритма.**<br>
<ol>
    <li>1. Задаётся некий фиксированный параметр $k$ для рассматриваемого числа $n$. Выполняется вычисление "вспомогательных чисел":</li>
    $$ d = 2^k, \quad q_1 = [\frac{n}{d}], \quad r_1 = n \; mod \; d \implies n = q_1d + r_1 \; (0 \leq r_1 < d) \\ q_2 = [\frac{q_1}{d}], \quad r_2 = q_1 mod \; d \implies q_1 = q_2d + r_2 \; (0 \leq r_2 < d)$$
    <li>2. Данная процедура продолжается до тех пор, пока не появится такое $q_s < d$. Следовательно $q_{s-1} = d_sd + r_s.$</li>
    <li>3. Таким образом, $n$ имеет вид:</li>
    $$ n = 2^kq_1 + r_1 = 2^k(2^kq_2 + r_2) + r_1 = \dots = 2^k(2^k(\dots(2^kq_s + r_s)\dots)+ r_2) + r_1 $$
    </ol>


#### Звёздные цепочки.<br>
**Определение.**<br>
Звёздной цепочкой называется аддитивная цепочка, включающая в себя только звёздные шаги.<br>
<ul>
    <li>Длина звездной цепочки $l^*(n)$.</li>
    <li>Шаг звёздной цепочки $a_i = a_{i-1} + a_k \; \forall k < i$.</li>
    <li>Очевидно, что $l(n) \leq l^*(n)$.</li>
    </ul>
    
#### Вектор индексов.<br>
**Определение**<br>
Пусть дана звездная цепочка длины $m-1$ вида $1 = a_1, a_2, \dots, a_m$. Для каждой звездной цепочки существует вектор индексов $r = \{r_1, r_2, \dots, r\}$ длины $m-1$ такой, что
$$ r_i = \{z:1\leq z\leq i\},\quad a_i = a_{i-1} + a_{r_{i-1}},\quad 2\leq i\leq m-1$$
<ul>
    <li>Первый элемент вектора $r_1$ - всегда $1$, второй $r_2 = \{1, 2\}$ и так далее. Последний элемент $1, \dots, m_1$.</li>
    <li>Наибольшая звездная цепочка с соответствующим ей вектором индексов имеют вид:</li>
    $a = \{1, 2, 4, \dots, 2^{m-1}\}$<br>
    $r = \{1, 2, 3, \dots, m-1\}$<br>
    <li>Наименьшая звездная цепочка с соответствующим ей вектором индексов имеют вид:</li>
    $a = \{1, 2, 4, \dots, m\}$<br>
    $r = \{1, 1, 1, \dots, 1\}$<br>
    <li>#$ \mathcal{A}^*(m-1) = (m-1)! $</li>
    </ul>
    
**Алгоритм дробления вектора индексов.**<br>
**Сравнимость двух векторов.**<br>
Даны 2 вектора индексов равной длины $\{r_1, r_2, \dots, r_m\}$, $\{\tilde{r}_1, \tilde{r}_2, \dots, \tilde{r}_m\}$. Тогда $r \succ \tilde{r}$, если $r_1 = \tilde{r}_1 = 1, r_2 = \tilde{r}_2, \dots, r_{i-1} = \tilde{r}_{i-1}$, причём $r_i > \tilde{r}_i$.

**Анализ вектора индексов.**
<ul>
    <li>При уменьшении последнего элемента вектора индексов с $m-1$ до $1$ последний элемент звёздной цепочки $a_m$ монотонно убывает.</li>
    <li>Таким образом, можно уменьшать не весь вектор сразу, а, например, половину.</li>
    <li>Рассмотрим вектор индексов вида:<br>
    $\{r_1, r_2, \dots, r_q\}\cup\{\rho_{q+1}, \rho_{q+2}, \dots, \rho_m\}$<br>
    Назовём левую часть фиксированной, правую меняющейся. Таких наборов $\frac{m!}{q!}.$</li>
    <li>$a_{min} = a_{q+1} + m - q$ при $\{r_1, r_2, \dots, r_q\}\cup\{1, 1, \dots, 1\}$<br>
        $a_{max} = a_{q+1} \cdot 2^{m-q}$ при $\{r_1, r_2, \dots, r_q\}\cup\{q+1, q+2, \dots, m\}$</li>
    </ul>
    
**Алгоритм дробления вектора индексов**
<ol>
    <li>Запускаем внешний цикл по длинам цепочек: $\underline{l}(n) \leq m \leq \bar{l}(n)$. Выбирается число $q \in \mathbb{N}$ - индекс дроблений $q = \frac{m}{2}.$</li>
    <li>Запускаем внутренний цикл перебора всех $\{r_1, r_2, \dots, r_q\}$ ($q!$ шагов). На каждом шаге при фиксированной части вычисляем $a_{min}$ и $a_{max}$ и строим цепочку:</li>
        <ol>
            <li>$a_m = n$ - задача решена.</li>
            <li>$n \notin [a_{min}, a_{max}]$, то переход к следующему набору $\{r_1, r_2, \dots, r_q\}$.</li>
            <li>$n \in [a_{min}, a_{max}]$, то организуем внутренний цикл перебора меняющейся части $\{\rho_{q+1}, \rho_{q+2}, \dots, \rho_m\}$. Таких набор $\frac{m!}{q!}$:</li>
            <ol>
                <li>Если $a_m = n$ - задача решена.</li>
                <li>Если таких векторов не оказалось, то переходим к следующему(по введённой упорядоченности) фиксированной части $\{r_1, r_2, \dots, r_q\}$.</li>
                </ol>
            </ol>
    <li>Если все наборы фиксированной длины исчерпаны, то увеличиваем их длину во внешнем цикле.</li>
    </ol>

#### Гипотеза Шольца-Брауэра.<br>
$$ l(2^n - 1) \leq l(n) + n - 1 $$
<ul>
    <li>Доказана для звёздных цепочек: $ l^*(2^n - 1) \leq l^*(n) + n - 1 $.</li>
    <li>Гипотеза справедлива дл всех $n < 578496$.</li>
    <li>Равенство выполняется для $1 \leq n \leq 64$.</li>
    </ul

## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы

### 1. Реализация алгоритма Брауэра.

Реализуем алгоритм Брауэра для вычисления приближенных аддитивных цепочек для различных чисел при варьировании параметра $k$.

Написана функция $Brouer$, на вход которой поступает число $n$ и фиксированный параметр $k$. С помощью цикла $for$ происходит начальное заполнение цепочки $chain$ числами от $1$ до $2^k - 1$. Далее составляется массив "вспомогательных чисел" $points$, которые необходимо получить в цепочке. Данный массив получается путем деления нацело числа $n$. С помощью цикла $for$ совершается проход по элементам массива $points$. Если элемент, лежащий в $points$ можно получить сложением двух чисел в цепочке, то добавляем $element$, иначе добавляем в цепочку умноженный на два последний элемент цепочки и повторяем действия до тех пор пока последний элемент цепочке не будет равет $n$.

In [7]:
def Brauer(n, k):
    d = 2 ** k
    chain = []
    points = [n]
    
    #Если 2^k больше исходного числа, то степень 2-ки уменьшается  
    if d > n:
        while d > n:
            d //= 2
    
    #Заполнение цепочки начальными элементами от 1 до 2^k - 1
    for i in range(1, d):
        chain.append(i)
        
    #Составление массива "вспомогательных чисел", которые необходимо получить в цепочке    
    tmp_n = n    
    while tmp_n > d:
        tmp_n = tmp_n // d
        points.append(tmp_n)
    points.reverse()
    
    #Если элемент, лежащий в points можно получить сложением двух элементов в цепочке, то добавляем элемент
    #Иначе умножаем на 2 последний элемент цепочки
    for element in points:
        if element < d:
            chain.append(element * 2)
        else:
            while chain[-1] != element:
                if element - chain[-1] in chain:
                    chain.append(element)
                else:
                    chain.append(chain[-1] * 2)
    #Избавление от повторений
    return sorted(list(set(chain)))

Опишем функцию $print$_$Brauer$, которая выводит на экран входные параметры $n$ и $k$, аддитивную цепочку и ее длину.

In [62]:
def print_Brauer(n, k):
    chain = Brauer(n, k)
    print("Число:", n)
    print("Фиксированный параметр:", k)
    print(chain)
    print("Длина аддитивной цепочки: ", len(chain))
    print()

Проверим работоспособность вышенаписанного алгоритма на числах $111, 563, 1077, 9781, 11574 $, варьируя параметр $k$ от $2$ до $4$:

In [78]:
test_array = [111, 563, 1077, 9781, 11574]
for elem in test_array:
        for k in range(2,5):
            print_Brauer(elem, k)

Число: 111
Фиксированный параметр: 2
[1, 2, 3, 4, 6, 12, 24, 27, 54, 108, 111]
Длина аддитивной цепочки:  11

Число: 111
Фиксированный параметр: 3
[1, 2, 3, 4, 5, 6, 7, 8, 13, 26, 52, 104, 111]
Длина аддитивной цепочки:  13

Число: 111
Фиксированный параметр: 4
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 24, 48, 96, 111]
Длина аддитивной цепочки:  19

Число: 563
Фиксированный параметр: 2
[1, 2, 3, 4, 8, 16, 32, 35, 70, 140, 280, 560, 563]
Длина аддитивной цепочки:  13

Число: 563
Фиксированный параметр: 3
[1, 2, 3, 4, 5, 6, 7, 8, 16, 32, 64, 70, 140, 280, 560, 563]
Длина аддитивной цепочки:  16

Число: 563
Фиксированный параметр: 4
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 32, 35, 70, 140, 280, 560, 563]
Длина аддитивной цепочки:  23

Число: 1077
Фиксированный параметр: 2
[1, 2, 3, 4, 8, 16, 32, 64, 67, 134, 268, 269, 538, 1076, 1077]
Длина аддитивной цепочки:  15

Число: 1077
Фиксированный параметр: 3
[1, 2, 3, 4, 5, 6, 7, 8, 16, 32, 64, 128, 134, 268, 536, 1072,

Представим тестирование в виде таблицы: <br>
(Параметры минимальных цепочек взяты с сайта: http://wwwhomes.uni-bielefeld.de/achim/addition_chain.html)

| n | k | Аддитивная цепочка | Длина цепочки | Минимальная цепочка | Длина минимальной цепочки |
|:---:|:---:|:---:|:---:|:---:|:---:|
| 111 | 2 | 1, 2, 3, 4, 6, 12, 24, 27, 54, 108, 111 | 11 | 1, 2, 3, 6, 12, 24, 27, 54, 108, 111 | 10 |
| 111 | 3 | 1, 2, 3, 4, 5, 6, 7, 8, 13, 26, 52, 104, 111 | 13 | 1, 2, 3, 6, 12, 24, 27, 54, 108, 111 | 10 |
| 111 | 4 | 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 24, 48, 96, 111 | 19 | 1, 2, 3, 6, 12, 24, 27, 54, 108, 111 | 10 |
| 563 | 2 | 1, 2, 3, 4, 8, 16, 32, 35, 70, 140, 280, 560, 563 | 13 | 1, 2, 4, 8, 9, 17, 34, 68, 136, 272, 281, 562, 563 | 13 |
| 563 | 3 | 1, 2, 3, 4, 5, 6, 7, 8, 16, 32, 64, 70, 140, 280, 560, 563 | 16 | 1, 2, 4, 8, 9, 17, 34, 68, 136, 272, 281, 562, 563 | 13 |
| 563 | 4 | 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 32, 35, 70, 140, 280, 560, 563 | 23 | 1, 2, 4, 8, 9, 17, 34, 68, 136, 272, 281, 562, 563 | 13 |
| 1077 | 2 | 1, 2, 3, 4, 8, 16, 32, 64, 67, 134, 268, 269, 538, 1076, 1077 | 15 | 1, 2, 3, 5, 8, 16, 32, 64, 67, 134, 268, 536, 1072, 1077 | 14 |
| 1077 | 3 | 1, 2, 3, 4, 5, 6, 7, 8, 16, 32, 64, 128, 134, 268, 536, 1072, 1077 | 17 | 1, 2, 3, 5, 8, 16, 32, 64, 67, 134, 268, 536, 1072, 1077 | 14 |
| 1077 | 4 | 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 32, 64, 67, 134, 268, 536, 1072, 1077 | 24 | 1, 2, 3, 5, 8, 16, 32, 64, 67, 134, 268, 536, 1072, 1077 | 14 |
| 9781 | 2 | 1, 2, 3, 4, 8, 9, 18, 36, 38, 76, 152, 304, 608, 611, 1222, 2444, 2445, 4890, 9780, 9781 | 20 | 1, 2, 3, 6, 12, 13, 19, 38, 76, 152, 304, 608, 1216, 2432, 2445, 4890, 9780, 9781 | 18 |
| 9781 | 3 | 1, 2, 3, 4, 5, 6, 7, 8, 16, 19, 38, 76, 152, 304, 608, 1216, 1222, 2444, 4888, 9776, 9781 | 21 | 1, 2, 3, 6, 12, 13, 19, 38, 76, 152, 304, 608, 1216, 2432, 2445, 4890, 9780, 9781 | 18 |
| 9781 | 4 | 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 32, 38, 76, 152, 304, 608, 611, 1222, 2444, 4888, 9776, 9781 | 28 | 1, 2, 3, 6, 12, 13, 19, 38, 76, 152, 304, 608, 1216, 2432, 2445, 4890, 9780, 9781 | 18 |
| 11574 | 2 | 1, 2, 3, 4, 8, 11, 22, 44, 45, 90, 180, 360, 720, 723, 1446, 2892, 2893, 5786, 11572, 11574 | 20 | 1, 2, 4, 8, 9, 18, 27, 45, 90, 180, 360, 720, 1440, 2880, 5760, 5787, 11574 | 17 |
| 11574 | 3 | 1, 2, 3, 4, 5, 6, 7, 8, 16, 22, 44, 88, 176, 180, 360, 720, 1440, 1446, 2892, 5784, 11568, 11574 | 22 | 1, 2, 4, 8, 9, 18, 27, 45, 90, 180, 360, 720, 1440, 2880, 5760, 5787, 11574 | 17 |
| 11574 | 4 | 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 32, 45, 90, 180, 360, 720, 723, 1446, 2892, 5784, 11568, 11574 | 28 | 1, 2, 4, 8, 9, 18, 27, 45, 90, 180, 360, 720, 1440, 2880, 5760, 5787, 11574 | 17 |

#### Выводы.

Алгоритм Брауэра достаточно быстро находит аддитивную цепочку. Но алгоритм не всегда работает оптимально. Это связано с тем, что в алгоритме дополнительно вычисляются "вспомогательные числа", которые могут быть не использованы. После тестирования мы убедились, что длины минимальных цепочек и длины цепочек, полученных с помощью алгоритма, не совпадают.

### 2. Реализация алгоритма дробления вектора индексов

Реализуем алгоритм дробления вектора индексов для нахождения минимальной звездной цепочки для заданного числа. <br>
Функция **create_chain(vector)** - данная функция на вход получает вектор, по которому составляет аддитивную цепочку, которую впоследствие возвращает. <br>
Функция **create_next_set(vector, q)** - данная функция принимает на вход вектор и текущий аргумент q. Составляет новый вектор и возвращает его.<br>
Функция **split_vector(n)** - данная функция является основной в алгоритме дробления вектора индексов. Работает по принципу, описанному выше в основных теоретических положениях.<br>

In [2]:
import time

# Получаем из вектора индексов аддитивную цепочку
def create_chain(vector):
    chain = [1]
    for i in vector:
        chain.append(chain[-1] + chain[i - 1])
    return chain

# Возвращает следующий набор
def create_next_set(vector, q):
    if (vector == [1 for i in range(len(vector))]):
        return [0]
    for i in range(len(vector) - 1, -1, -1):
        if (vector[i] == 1):
            vector[i] = q + i + 1
        else:
            vector[i] -= 1
            return vector
    return vector
    
def split_vector(n, fixed_m = None):
    # Внешний цикл l_(n) <= m <= _l(n)
    for m in range(int(log(n, 2)), int(log(n, 2)) + bin(n).count('1')):
        q = m // 2 
        r_q = [i for i in range(1, q + 1)]
        r_m = [q + i + 1 for i in range(m - q)]
        # Внутренний цикл, перебо всех наборов r_q (пока не все 1 в векторе)
        while (sum(r_q) >= q ):
            chain = create_chain(r_q + r_m)
            # a_m = n
            if(chain[-1] == n):
                return chain
            # Если n не принадлежит [a_min, a_max]
            if n < (chain[q] + m - q) or chain[q] * 2**(m-q) < n:
                # Переходим к след. набору фикс. часть
                r_q = create_next_set(r_q, 0)
                continue
            # n принадлежит [a_min, a_max], начинаем перебор r_m  (пока не все 1 в векторе)
            while (sum(r_m) >= (m-q)):
                chain = create_chain(r_q + r_m)
                # a_m = n
                if (n == chain[-1]):
                    return chain
                # Переходим к след. набору
                r_m = create_next_set(r_m, q)
            # Переходим к след. набору фикс. часть
            r_q = create_next_set(r_q, 0)
            r_m = [q + i + 1 for i in range(m - q)]
    
# Выводит результат
def print_result_split(n):
    print('n =', n)
    t1 = time.time()            
    chain = split_vector(n)
    t2 = time.time() - t1
    print(chain)
    print('Длина цепочки:', len(chain))
    print("Время работы алгоримта {0:.5f} s".format(t2), '\n')
    

Протестируем алгоритм на следующих значениях: $1024, 1027, 1036, 1040, 1056, 1152$. В выводе укажем сколько времени потребовалось на поиск цепочки, какая цепочка получилась и ее длина.

In [4]:
test_array_for_split_vector = [1024, 1027, 1036, 1040, 1056, 1152]
for element in test_array_for_split_vector:
    print_result_split(element)

n = 1024
[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]
Длина цепочки: 11
Время работы алгоримта 0.00017 s 

n = 1027
[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1026, 1027]
Длина цепочки: 13
Время работы алгоримта 55.99493 s 

n = 1036
[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1032, 1036]
Длина цепочки: 13
Время работы алгоримта 56.08322 s 

n = 1040
[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1040]
Длина цепочки: 12
Время работы алгоримта 0.00191 s 

n = 1056
[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1056]
Длина цепочки: 12
Время работы алгоримта 0.00161 s 

n = 1152
[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1152]
Длина цепочки: 12
Время работы алгоримта 0.00158 s 



Для тех же чисел запустим алгоритм Брауэра с параметром $k$ равным оптимальному значению $2$. В выводе также укажем сколько времени потребовалось на поиск цепочки, какая цепочка получилась и ее длина. Для этого немного перепишем функцию вывода $print$_$Brauer(n, k)$.

In [ ]:
def print_Brauer_with_time(n, k):
    t1 = time.time()
    chain = Brauer(n, k)
    t2 = time.time() - t1
    print("Число:", n)
    print("Фиксированный параметр:", k)
    print(chain)
    print("Длина аддитивной цепочки: ", len(chain))
    print("Время работы алгоримта {0:.5f} s".format(t2), '\n')

In [8]:
for element in test_array_for_split_vector:
    print_Brauer_with_time(element, 2)

Число: 1024
Фиксированный параметр: 2
[1, 2, 3, 4, 8, 16, 32, 64, 128, 256, 512, 1024]
Длина аддитивной цепочки:  12
Время работы алгоримта 0.00010 s 

Число: 1027
Фиксированный параметр: 2
[1, 2, 3, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1027]
Длина аддитивной цепочки:  13
Время работы алгоримта 0.00003 s 

Число: 1036
Фиксированный параметр: 2
[1, 2, 3, 4, 8, 16, 32, 64, 128, 256, 259, 518, 1036]
Длина аддитивной цепочки:  13
Время работы алгоримта 0.00004 s 

Число: 1040
Фиксированный параметр: 2
[1, 2, 3, 4, 8, 16, 32, 64, 65, 130, 260, 520, 1040]
Длина аддитивной цепочки:  13
Время работы алгоримта 0.00006 s 

Число: 1056
Фиксированный параметр: 2
[1, 2, 3, 4, 8, 16, 32, 64, 66, 132, 264, 528, 1056]
Длина аддитивной цепочки:  13
Время работы алгоримта 0.00003 s 

Число: 1152
Фиксированный параметр: 2
[1, 2, 3, 4, 8, 16, 18, 36, 72, 144, 288, 576, 1152]
Длина аддитивной цепочки:  13
Время работы алгоримта 0.00003 s 



Результаты тестирования алгоритма дробления вектора индекса и алгоритма Брауэра сведем в одну таблицу для наглядности.

| n | Алгоритм дробления вектора индекса | Длина цепочки | Время работы | Алгоритм Брауэра | Длина цепочки | Время работы |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| 1024 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] | 11 | 0.00017 s | [1, 2, 3, 4, 8, 16, 32, 64, 128, 256, 512, 1024] | 12 | 0.00010 s |
| 1027 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1026, 1027] | 13 | 55.99493 s | [1, 2, 3, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1027] | 13 | 0.00003 s |
| 1036 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1032, 1036] | 13 | 56.08322 s | [1, 2, 3, 4, 8, 16, 32, 64, 128, 256, 259, 518, 1036] | 13 | 0.00004 s |
| 1040 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1040] | 12 | 0.00191 s | [1, 2, 3, 4, 8, 16, 32, 64, 65, 130, 260, 520, 1040] | 13 | 0.00006 s |
| 1056 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1056] | 12 | 0.00161 s | [1, 2, 3, 4, 8, 16, 32, 64, 66, 132, 264, 528, 1056] | 13 | 0.00003 s |
| 1152 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1152] | 12 | 0.00158 s | [1, 2, 3, 4, 8, 16, 18, 36, 72, 144, 288, 576, 1152] | 13 | 0.00003  |

#### Выводы.
Реализован алгоритм дробления вектора индексов. Проведено тестирование на $n > 1000$ алгоритма дробления вектора индексов и алгоритма Брауэра. По результатам тестирования можно сделать вывод, что алгоритм дробления вектора индекса находит минимальную аддитивную цепочку, но за достаточно большое время, так как алгоритм реализован на основе полного перебора. Алгоритм Брауэра выигрывает по времени алгоритм дробления вектора индексов, но при этом проигрывает ему в точности и оптимальности.

### 3. Проверка гипотезы Шольца-Брауэра

Проверим гипотезу Шольца-Брауэра для всех натуральных чисел $1 \leq n \leq 12$ на алгоритме дробления вектора индексов, которая гласит, что для этих чисел должно выполняться неравенство: $$l(2^n - 1) \leq l(n) + n - 1$$

In [4]:
def testing_hypothesis():
    is_hypothesis = True
    for n in range(1, 13):
        first_chain_len = len(split_vector(n)) + n - 1
        second_chain_len = len(split_vector(2 ** n - 1))
        if second_chain_len > first_chain_len:
            print("Гипотеза Шольца-Брауэра НЕВЕРНА для", first_chain_len, second_chain_len)
            is_hypothesis = False
        else:
            print("l(n) + n - 1 =", first_chain_len)
            print("l(2^n - 1) =", second_chain_len)
            print("Гипотеза Шольца-Брауэра ВЕРНА для", n)
            print()
    if is_hypothesis:
        print("Гипотеза Шольца-Брауэра ВЕРНА для всех заданных n")

testing_hypothesis()

l(n) + n - 1 = 1
l(2^n - 1) = 1
Гипотеза Шольца-Брауэра ВЕРНА для 1

l(n) + n - 1 = 3
l(2^n - 1) = 3
Гипотеза Шольца-Брауэра ВЕРНА для 2

l(n) + n - 1 = 5
l(2^n - 1) = 5
Гипотеза Шольца-Брауэра ВЕРНА для 3

l(n) + n - 1 = 6
l(2^n - 1) = 6
Гипотеза Шольца-Брауэра ВЕРНА для 4

l(n) + n - 1 = 8
l(2^n - 1) = 8
Гипотеза Шольца-Брауэра ВЕРНА для 5

l(n) + n - 1 = 9
l(2^n - 1) = 9
Гипотеза Шольца-Брауэра ВЕРНА для 6

l(n) + n - 1 = 11
l(2^n - 1) = 11
Гипотеза Шольца-Брауэра ВЕРНА для 7

l(n) + n - 1 = 11
l(2^n - 1) = 11
Гипотеза Шольца-Брауэра ВЕРНА для 8

l(n) + n - 1 = 13
l(2^n - 1) = 13
Гипотеза Шольца-Брауэра ВЕРНА для 9

l(n) + n - 1 = 14
l(2^n - 1) = 14
Гипотеза Шольца-Брауэра ВЕРНА для 10

l(n) + n - 1 = 16
l(2^n - 1) = 16
Гипотеза Шольца-Брауэра ВЕРНА для 11

l(n) + n - 1 = 16
l(2^n - 1) = 16
Гипотеза Шольца-Брауэра ВЕРНА для 12

Гипотеза Шольца-Брауэра ВЕРНА для всех заданных n


Оформим результат тестирования в виде таблицы:

| n | l(2^n - 1) | l(n) + n - 1 | Гипотеза |
|:---:|:---:|:---:|:---:|
| 1 | 1 | 1 | Верна |
| 2 | 3 | 3 | Верна |
| 3 | 5 | 5 | Верна |
| 4 | 6 | 6 | Верна |
| 5 | 8 | 8 | Верна |
| 6 | 9 | 9 | Верна |
| 7 | 11 | 11 | Верна |
| 8 | 11 | 11 | Верна |
| 9 | 13 | 13 | Верна |
| 10 | 14 | 14 | Верна |
| 11 | 16 | 16 | Верна |
| 12 | 16 | 16 | Верна |

#### Выводы.


Гипотеза Шольца-Брауэра подтверждена для чисел $1 \leq n \leq 12$ на алгоритме дробления вектора индексов.<br>
Для подтверждения гипотезы потребовалось большое количество времени, из чего следует вывод, что алгоритм дробления вектора индексов требует совершенствования и оптимизации.

## Выводы


В ходе практической работы реализован приближенный алгоритм нахождения минимальной цепочки, называемый алгоритмом Брауэра, и реализован точный алгоритм нахождения минимальной цепочки, называемый алгоритмом дробления вектора индексов.<br>
На практике установлена работоспособность данных алгоритмов. Выяснено, что алгоритм дробления вектора индексов работает точно, но очень долго, так как он основан на полном переборе. Алгоритм Брауэра работает достаточно быстро, но при этом не всегда составляет минимальную аддитивную цепочку.<br>
Также проверена гипотеза Шольца-Брауэра на числах $1 \leq n \leq 12$ на алгоритме дробления векторов индексов.